### PR TITLE
Add filtering of group name to autocomplete for transactions 

### PR DESCRIPTION
**Pull request by Crazypkr1099**
Reference #1681 

I am anticipating criticism, so I am all ears. If you have a better place to put this option, please let me know :) I just really need this option.

### CODE REVIEW

**Comment by youngcw** [comment]
The theme button on the top of the page was removed recently. It shows up in development but not production for color testing.

**Comment by Crazypkr1099** [comment]
> The theme button on the top of the page was removed recently. It shows up in development but not production for color testing.

Why was this removed???? it was a great place for it.

**Comment by carkom** [comment]
It looks pretty good. Can you remove all the code for the settings option as that is no longer relevant?

**Review by psybers — CHANGES_REQUESTED** [review]

**Comment by psybers** [comment]
So I mentioned this in the Discord and thought I would bring it up again.  If I search for 'Bills' (using the test dataset) I see the category group and all categories.  If I keep typing 'Bills cell' then it goes to nothing.  If I want 'cell' I would have to up/down to select it, or clear the search and type 'cell'.  I think we can do better here.  

```js
  function customSort(obj, value) {
    const name = obj.name.toLowerCase();
    const groupName = obj.group ? obj.group.name.toLowerCase() : '';
    const longName = `${groupName} ${name}`;
    value = value.toLowerCase();

    if (obj.id === 'split') {
      return -2;
    }
    if (name.includes(value)) {
      return -1;
    }
    if (groupName.includes(value)) {
      return 0;
    }
    if (obj.group && longName.includes(value)) {
      return 0;
    }
    return 1;
  }
```

and then

```js
          return (
            suggestion.id === 'split' ||
            suggestion.group?.name
              .toLowerCase()
              .includes(value.toLowerCase()) ||
            (suggestion.group?.name + ' ' + suggestion.name)
              .toLowerCase()
              .includes(value.toLowerCase()) ||
            defaultFilterSuggestion(suggestion, value)
          );
```

The idea is to search a string that is the category name followed by the category.  My quick test seems like it works well but I am open to suggestions.

image

**Comment by Crazypkr1099** [comment]
> So I mentioned this in the Discord and thought I would bring it up again.  If I search for 'Bills' (using the test dataset) I see the category group and all categories.  If I keep typing 'Bills cell' then it goes to nothing.  If I want 'cell' I would have to up/down to select it, or clear the search and type 'cell'.  I think we can do better here.  
> 
> ```js
>   function customSort(obj, value) {
>     const name = obj.name.toLowerCase();
>     const groupName = obj.group ? obj.group.name.toLowerCase() : '';
>     const longName = `${groupName} ${name}`;
>     value = value.toLowerCase();
> 
>     if (obj.id === 'split') {
>       return -2;
>     }
>     if (name.includes(value)) {
>       return -1;
>     }
>     if (groupName.includes(value)) {
>       return 0;
>     }
>     if (obj.group && longName.includes(value)) {
>       return 0;
>     }
>     return 1;
>   }
> ```
> 
> and then
> 
> ```js
>           return (
>             suggestion.id === 'split' ||
>             suggestion.group?.name
>               .toLowerCase()
>               .includes(value.toLowerCase()) ||
>             (suggestion.group?.name + ' ' + suggestion.name)
>               .toLowerCase()
>               .includes(value.toLowerCase()) ||
>             defaultFilterSuggestion(suggestion, value)
>           );
> ```
> 
> The idea is to search a string that is the category name followed by the category.  My quick test seems like it works well but I am open to suggestions.
> 
> image
> 

I'll add this fix in later tonight or tomorrow

**Comment by yoyotogblo** [comment]
Thanks for working on this!

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (psybers) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:227]:
```suggestion
      renderItems={(items, getItemProps, highlightedIndex) => (
        <CategoryList
          items={items}
          embedded={embedded}
          getItemProps={getItemProps}
          highlightedIndex={highlightedIndex}
          renderSplitTransactionButton={renderSplitTransactionButton}
          renderCategoryItemGroupHeader={renderCategoryItemGroupHeader}
          renderCategoryItem={renderCategoryItem}
          showHiddenItems={showHiddenCategories}
          showBalances={showBalances}
        />
      )}
```

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (psybers) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:125]: Did this fix a lint issue? If not, I would revert.
```suggestion
                      item.hidden && { color: theme.pageTextSubdued }),
```

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:214]: **🥜 nitpick:** ‏another set of unnecessary changes here

File: packages/desktop-client/src/components/autocomplete/Autocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/Autocomplete.tsx:324]: **🔨 warning:** ‏a couple of notes on the sorting logic:

1. the generic autocomplete component is not a good place for the custom sorting logic that is applicable only to category autocomplete - a better location would be in `CategoryAutocomplete`
2. instead of using `.sort` (which mutates the array in-place thus causing extra re-renders in the component tree) -  use `.toSorted`
3. remember to `useMemo` the result - a small change to improve the perf.
4. missing TS types for the custom sort util; plus - if it's a generic - it should be moved _outside_ the component - thus reducing the amount of re-renders

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Usability works as expected. Thanks for working on this! 👍

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:235]: **💬 suggestion:** ‏`sort` mutates the input (sorts in place). Lets use [`toSorted`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSorted) to do the same thing, but instead copy to a new object.

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:219]: **🔨 warning:** ‏manual type coercions more often than not lead to subtle bugs down-the-line. Do we event need to use the `Suggestion` type here? AFAIK the `suggestions` (variable) underlying type is `CategoryAutocompleteItem`. So we could just use that.

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:238]: **🔨 warning:** ‏same type coercion issue here.

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:211]: **💬 suggestion:** ‏good idea to use memoization! However, we should use `useCallback` here instead.

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:149]: **🔨 warning:** ‏we should treat variables as immutable. Otherwise this has the potential to introduce subtle and hard-to-debug bugs.

```suggestion
  const lowerCaseValue = value.toLowerCase();
```

**Review by psybers — APPROVED** [review]

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (psybers) [reply packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:149]: Is this even needed? The calls to `customSort` below seem to pass in lowercased values anyway.

Comment (psybers) [reply packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:149]: A safer approach (in case someone adds more calls to this function) would be to lowercase in here as it does, and just drop the lowercase at the call sites below.

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:221]: Ok, looks like `toSorted` isn't available in the version of TS (I assume?) we're using. Sorry about that. Lets bring back `.sort()`

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:209]: **💬 suggestion:** ‏no need for this type coercion anymore
```suggestion
        .filter(suggestion => {
```

File: packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/autocomplete/CategoryAutocomplete.tsx:227]: **🥜 nitpick:** ‏
```suggestion
    [],
```

**Review by MatissJanis — APPROVED** [review]
LGTM! Thanks for your hard work on this!

(will wait for another maintainer to review before merging)

**Review by psybers — APPROVED** [review]

**Review by youngcw — APPROVED** [review]
Looks good

**Pull request merged**
